src/app/qgsversionmigration.cpp

```diff
@@ -0,0 +1,222 @@
+{
+  QgsError error;
+  QString oldHome = QStringLiteral( "%1/.qgis2" ).arg( QDir::homePath() );
+  QString oldStyleFile = QStringLiteral( "%1/symbology-ng-style.db" ).arg( oldHome );
+  QgsDebugMsg( QString( "OLD STYLE FILE %1" ).arg( oldStyleFile ) );
+  QSqlDatabase db = QSqlDatabase::addDatabase( "QSQLITE", "migration" );
+  db.setDatabaseName( oldStyleFile );
+  if ( !db.open() )
+  {
+    error.append( db.lastError().text() );
+    QgsDebugMsg( db.lastError().text() );
+    return error;
+  }
+
+  QSqlQuery query( db );
+  QSqlQuery tagQuery( "SELECT name FROM tag"
+                      "JOIN tagmap ON tagmap.tag_id = tag.id"
+                      "WHERE tagmap.symbol_id = :symbol_id", db );
+
+  QgsStyle *style = QgsStyle::defaultStyle();
+  if ( query.exec( "SELECT id, name, xml FROM symbol" ) )
+  {
+    while ( query.next() )
+    {
+      QString symbol_id = query.value( 0 ).toString();
+      QString name = query.value( 1 ).toString();
+      QString xml = query.value( 2 ).toString();
+      QDomDocument doc;
+      if ( !doc.setContent( xml ) )
+      {
+        QgsDebugMsg( "Cannot open symbol " + name );
+        continue;
+      }
+
+      tagQuery.bindValue( ":symbol_id", symbol_id );
+
+      QStringList tags;
+      if ( tagQuery.exec() )
+      {
+        while ( query.next() )
+        {
+          QString tagname = query.value( 0 ).toString();
+          tags << tagname;
+        }
+      }
+
+      QDomElement symElement = doc.documentElement();
+      QgsDebugMsg( QString( "MIGRATION: Importing %1" ).arg( name ) );
+      QgsSymbol *symbol = QgsSymbolLayerUtils::loadSymbol( symElement, QgsReadWriteContext() );
+      tags << "QGIS 2";
+      if ( style->symbolId( name ) == 0 )
+      {
+        style->saveSymbol( name, symbol, false, tags );
+      }
+    }
+  }
+
+  QgsDebugMsg( oldStyleFile );
+  return error;
+}
+
+QgsError Qgs2To3Migration::migrateSettings()
+{
+  QgsError error;
+
+  QgsSettings newSettings;
+
+  // The platform default location for the settings from 2.x
+  QSettings *oldSettings = new QSettings( "QGIS", "QGIS2" );
+
+  QFile inputFile( migrationFilePath() );
+  std::map<QString, QgsSettings::Section> sections;
+  sections["none"] = QgsSettings::NoSection;
+  sections["core"] = QgsSettings::Core;
+  sections["gui"] = QgsSettings::Gui;
+  sections["server"] = QgsSettings::Server;
+  sections["plugins"] = QgsSettings::Plugins;
+  sections["auth"] = QgsSettings::Auth;
+  sections["app"] = QgsSettings::App;
+  sections["providers"] = QgsSettings::Providers;
+  sections["misc"] = QgsSettings::Misc;
+
+  if ( inputFile.open( QIODevice::ReadOnly | QIODevice::Text ) )
+  {
+    QTextStream in( &inputFile );
+    while ( !in.atEnd() )
+    {
+      QString line = in.readLine();
+
+      if ( line.startsWith( "#" ) )
+        continue;
+
+      QStringList parts = line.split( ";" );
+
+      Q_ASSERT_X( parts.count() == 3, "QgsVersionMigration::migrateSettings()", "Can't split line in 3 parts." );
```

Comment:
for release builds, better put a line after this like:

    if ( parts.count() != 3)
      continue;